### DISCUSSION TITLE
How to "migrate" from existing read-only galleries to the new Libraries?

### DISCUSSION THREAD

**Discussion by alien75** [immich-app/immich#4240] [Q&A] [answered]
Hello, is there a suggested procedure to "migrate" from existing read-only galleries to the new Libraries?
Maybe I should simply remove the existing read-only galleries for all users, but how can I do it?
Thank you

**Comment by maltokyo** [comment]
I have the same question..  but there is a duplicate exact same discussion - maybe we could choose just one?
I have 115,000 photos already fully indexed etc, and dont want to do that process again, as it took 3 days to process everything in the "external gallery"?

https://github.com/immich-app/immich/discussions/4250

**Reply by davidecavestro** [reply]
I closed https://github.com/immich-app/immich/discussions/4250 as a duplicate: let's continue here.

**Reply by alien75** [reply]
Your contribution in #4250 with screens and query is useful: add it here please 🙂

**Reply by davidecavestro** [reply]
>Your contribution in https://github.com/immich-app/immich/discussions/4250 with screens and query is useful: add it here please 🙂

I started importing photos with [read-only gallery](https://documentation.immich.app/docs/features/read-only-gallery), then upgraded Immich and switched to [external library](https://documentation.immich.app/docs/features/libraries#external-libraries): I created an external library having the same import path as I originally set for the CLI import and did a rescan.

Now I suppose all assets have been duplicated
![image](https://github.com/immich-app/immich/assets/447917/6997db15-755f-4b69-8237-38432c5d15dd)
in fact if I run this query
```sql
SELECT
  COUNT(*), string_agg("id"::text, ', '), "originalPath"
FROM
  assets
GROUP BY
  "originalPath"
HAVING
  COUNT(*) > 1
ORDER BY
  COUNT(*) DESC
LIMIT 5;
```
I see a lot of assets sharing the same import path
![image](https://github.com/immich-app/immich/assets/447917/616cde23-5a05-4208-9fb3-0d25c16dfd96)

**Comment by etnoy** [comment]
There's no way right now, but we might look at in the future. In the meantime, you can keep your read-only imports as they still should work :)

**Reply by alien75** [reply]
Thanks.
Since I have another batch of 30k pictures I want to remove the existing [Read-only galleries](https://documentation.immich.app/docs/features/read-only-gallery) and upload old and new albums using the new [Libraries](https://documentation.immich.app/docs/features/libraries): can you please instruct on how to perform a clean removal of the read only galleries?

**Reply by FSchiltz** [reply]
Unfortunately read only imports are not equivalent to library, I would like to keep using the new librairies.
Read only doesn't works with RAW files and a lot of video.

**Reply by etnoy** [reply]
Are you familiar with SQL? If you run a SQL command against PostgreSQL you can instruct it to remove all imported assets. Take a backup first, though

**Reply by FSchiltz** [reply]
I marked them offline and not visible. They are not visible anymore and I hope the backend will see that and remove the thumbnails and transcode artifacts. 
Just deleting the row would have still left the artifacts orphan. 
Thanks, I didn't want to have to reimport 5TB of picture again.

**Comment by davidecavestro** [comment]
> There's no way right now, but we might look at in the future. In the meantime, you can keep your read-only imports as they still should work :)

So I removed the external library that lead to duplications.
Then I launched the import from CLI to sync new contents collected in the meantime by _rclone_, but I got bad requests complaining for lack of _libraryId_ param over a lot of files
`docker exec -it immich_microservices immich upload --key *** --server http://immich-server:3001 /Volume1/Photos/data_gphotos --recursive --import`

![image](https://github.com/immich-app/immich/assets/447917/6a4a13d5-87e6-4d15-84a1-e139d4bca402)

![image](https://github.com/immich-app/immich/assets/447917/602845f3-6b22-4cbc-832c-08b367c3e9fb)

To be honest I have no idea whether it happened for all files or just some of them (for which it happened potentially even during previous imports).

**Reply by FSchiltz** [reply]
You may need to update your cli by running `npm i -g immich` again

**Reply by davidecavestro** [reply]
Please note I've ran it from the immich microservices container itself. Even forcing the upgrade of CLI with npm it still sticks to 0.41.0 (the latest version as of now, bundled with immich v1.79.1)

**Reply by FSchiltz** [reply]
Strange,  I had the same issue as you that went away by updating the CLI and the container

**Reply by davidecavestro** [reply]
These are the versions I'm playing with
![image](https://github.com/immich-app/immich/assets/447917/b0252c06-fd07-4ccf-8c52-7d34eca28d59)

**Reply by jrasm91** [reply]
There's a bug on the server side they should be fixed next release.

**Reply by jrasm91** [reply]
https://github.com/immich-app/immich/pull/4224

**Comment by FSchiltz** [comment]
I deleted them in the database. But it does not remove the generated preview and transcoded video from disk .

First I found the UUID for my default library using 

```
SELECT l."id" AS library,  u."id" AS user
 FROM libraries AS l JOIN users AS u on u."id" = l."ownerId"
 WHERE l."type" = 'UPLOAD' 
 AND u."email" = 'test@test.com'
```

then I reviewed the change with 

```
SELECT
  "originalPath"
FROM
  assets
WHERE "libraryId" = '2e12b058-0ef8-4129-ae57-683dd9ee472c' AND "ownerId" = 'c0ec9624-c838-438e-a20a-8712fd85dd13';

```

then I just marked the assets as offline and not visible with 

```
UPDATE assets SET "isVisible" = false, "isOffline" = true
WHERE "libraryId" = '2e12b058-0ef8-4129-ae57-683dd9ee472c' AND "ownerId" = 'c0ec9624-c838-438e-a20a-8712fd85dd13';

```

**Reply by FSchiltz** [reply]
For anyone doing the same, remove your album first otherwise they will be in a broken state

**Comment by davidecavestro** [comment]
> I deleted them in the database. But it does not remove the generated preview and transcoded video from disk.

This is a quick and dirty way to rename any _.webp_ to _.webp.bak_ when its name doesn't match any asset.
So that you can review changes and then remove all unwanted stuff.
Please note I've initially left commented the `mv` command for safety purposes.

Please take a backup first in case you give it a try, and also adapt it to your env (path, container name, and so on)
The same could be done for jpeg thumbs and encoded videos.

```bash
EXTENSION=webp
THUMBS_DIR=/Volume1/docker/immich/photos/thumbs/53b1d4ae-9445-4273-9582-c8fbae49ee33
for THUMB in "${THUMBS_DIR}/"*.${EXTENSION}
do
#  echo "THUMB: ${THUMB}"
  ASSET_ID=$(basename -s .${EXTENSION} "${THUMB}")
#  echo "ASSET_ID: ${ASSET_ID}"
  FOUND=$(docker exec -it immich-database-1 psql -U postgres -h localhost immich -Atq -c "SELECT id FROM assets WHERE id='${ASSET_ID}'");
#  echo "FOUND: ${FOUND}"

  if [ -z "${FOUND}" ]; then
    echo "Asset not found for ${THUMB}"
#    mv "${THUMB}" "${THUMB}.bak" 
  fi
done
```

**Reply by davidecavestro** [reply]
Ouch, beware this way I've lost also face thumbs... I will regenerate

**Reply by davidecavestro** [reply]
Probably a Union between SELECT from assets and faces would do the trick... I'll check ASAP

**Reply by FSchiltz** [reply]
I'll keep my duplicates thumbnail until a cleaning is added to the app, it's not a lot of storage.
That might be a good idea, storing the face thumb in another folder than the image thumbnail

**Reply by davidecavestro** [reply]
> I'll keep my duplicates thumbnail until a cleaning is added to the app, it's not a lot of storage.
> That might be a good idea, storing the face thumb in another folder than the image thumbnail

Also keeping so many files directly within a single folderinstead of prefix-based  subfolders is a PITA, but I think there's a separate issue for that.

**Reply by jrasm91** [reply]
Yeah, and it has already been merged actually (#4112). It includes another job for migrating old thumbnails to the new path/format as well.

**Comment by andrejohansson** [comment]
I have a follow up questions regarding this.

For me, removing the read-only gallery setup and do a re-indexing is *acceptable* but I want to know how I should I approach it?

My current idea was:

- Add new mount volume to docker 

```yaml
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
     # Remove this existing
     - /mnt/media/mypics:/mnt/media/mypics:ro
     # Add this, same folder but new mount name
     - /mnt/media/mypics:/mnt/media/mypics-library:ro
```

- Run some command to detect that all images from read-only gallery have gone offline (which command, help here)
- Run remove offline command to delete read-only gallery images
- Add library using your [instructions](https://immich.app/docs/features/libraries) and point it to the new volume name `/mnt/media/mypics-library`

Follow up questions
- Does the approach seem viable?
- I guess I will loose my albums? 
    - Or will immich be inteliigent enough to detect identical hashes on images and keep my albums?
    - Can I take some other approach to keep my albums, or restore them afterwards?

**Reply by mattjmeier** [reply]
I wanted to add something to your list of questions...

I have recently started using Immich now that the external libraries are available. It is awesome! I'm seriously impressed by the UI and features.

Having started my libraries from scratch, there is, however, another issue that I think needs some consideration here (and also relates tangentially to some other issues, e.g., #4169, #2129, #1553, #2962). When backing up photos from mobile, hashes are only checked against the upload library. This means that - for several thousand photos on my device that I have already backed up in my external library - **they will appear as duplicates in the mobile timeline** (and, if I choose to back up those folders, they will also become actual duplicate assets on the server too).

This obviously is currently the intended operation, but I thought for anyone in a similar situation, it might be worth knowing about. Also, it seems like it would be really useful for the admin to decide which libraries should be included in duplicate hash detection.

It also kind of ties in to the discussion on photo "stacks" or "groups" (i.e., #2479). Would it be useful to automatically stack/group every photo that appears as duplicate hashes across multiple libraries? I really can't imagine a scenario where it is desirable to show duplicate photos in the timeline that have 100% identical content adjacent to each other.

Anyway: sorry if this isn't the most appropriate spot to ask, but I thought it was a sort of cross-cutting issue that relates mainly to the new libraries feature.

Thanks again to all the awesome developers who are making this project happen!

**Reply by etnoy** [reply]
This is an excellent question and I assure you that this is something I've thought about, too. I don't have a good solution (yet) but I hope to spend some time contemplating this and other aspects of external libraries that itch me as well.

Again, the big goal was to get libraries working. Now that they do we have a lot of polish to do and then we can look at new features like maybe something that tackles this use case (which I also have)

**Comment by davidecavestro** [comment, marked as answer]
It seems the relation between assets and libraries is simply backed by the _libraryId_ field of the _assets_ table.
![image](https://github.com/immich-app/immich/assets/447917/4fb3ac18-ee0d-42e6-8e05-c379aa540f68)

So I created a new library and set the same import path prevously used to "upload" assets to standard library (read-only gallery).
Then - since all my assets belong to a single library - I updated the libraryId of my assets to the one of the new lirary

`UPDATE assets SET "libraryId"='...';`

This could be refined adding a WHERE clause in case just a subset of assets should be moved.

This way all assets "magically moved" to the external library
![image](https://github.com/immich-app/immich/assets/447917/88626847-f7e7-4064-abf5-a75b416aea0b)

**Reply by etnoy** [reply]
Nice work. You might also want to set isExternal to true.

Please review your import paths so the assets aren't set to offline on the next refresh!

**Reply by davidecavestro** [reply]
So I issued this update, since all assets prevously imported using read-only gallery were under that path
`UPDATE assets SET "isExternal" = true WHERE "originalPath" LIKE '/Volume1/Photos/data_gphotos/%';`

**Reply by davidecavestro** [reply]
So far I see no side effects at all: it works like a charm!

**Reply by denniskns** [reply]
thanks for this solution, works great for me!

**Reply by alien75** [reply]
Excellent, thank you! Marked as answer.

**Reply by maltokyo** [reply]
Worked for me as well, thank you!
